# -*- coding: utf-8 -*-
from setuptools import setup

packages = \
['sequana_pipelines',
 'sequana_pipelines.multitax',
 'sequana_pipelines.multitax.data']

package_data = \
{'': ['*']}

install_requires = \
['click-completion>=0.5.2,<0.6.0',
 'sequana>=0.15.0',
 'sequana_pipetools>=0.16.1']

entry_points = \
{'console_scripts': ['sequana_multitax = sequana_pipelines.multitax.main:main']}

setup_kwargs = {
    'name': 'sequana-multitax',
    'version': '0.12.0',
    'description': 'A multi-sample and multi-databases taxonomic analysis using Kraken',
    'long_description': '\n.. image:: https://badge.fury.io/py/sequana-multitax.svg\n     :target: https://pypi.python.org/pypi/sequana_multitax\n\n\n.. image:: https://github.com/sequana/multitax/actions/workflows/main.yml/badge.svg\n   :target: https://github.com/sequana/multitax/actions/workflows/main.yaml \n\n.. image:: https://img.shields.io/badge/python-3.8%20%7C%203.9%20%7C3.10-blue.svg\n    :target: https://pypi.python.org/pypi/sequana\n    :alt: Python 3.8 | 3.9 | 3.10\n\n.. image:: http://joss.theoj.org/papers/10.21105/joss.00352/status.svg\n    :target: http://joss.theoj.org/papers/10.21105/joss.00352\n    :alt: JOSS (journal of open source software) DOI\n\nThis is is the **multitax** pipeline from the `Sequana <https://sequana.readthedocs.org>`_ project\n\n:Overview: Runs taxonomic analysis on a set of samples using sequana_taxonomy (kraken behing the scene)\n:Input: A set of Fastq files\n:Output: HTML report for each sample and a summary HTML report for all samples.\n:Status: Production\n:Citation: Cokelaer et al, (2017), ‘Sequana’: a Set of Snakemake NGS pipelines, Journal of Open Source Software, 2(16), 352, JOSS DOI doi:10.21105/joss.00352\n\n\nInstallation\n~~~~~~~~~~~~\n\nIf you already have all requirements, you can install the packages using pip::\n\n    pip install sequana_multitax --upgrade\n\n\nUsage\n~~~~~\n\nIn order to use this pipeline, you will need Kraken databases. Please see Kraken pages for help.\n\nWe will also need a taxonomic databases. In principle this pipeline will download the file in your home, in \n/home/user/.config/sequana/taxonomy.dat one for all. Note, that with singularity, this file will be downloaded locally in your working directory for each analysis,except if it is found in your home, in which case a simple copy is performed.\n\n::\n\n    sequana_multitax --help\n    sequana_multitax --input-directory DATAPATH  --databases toydb\n\nFor the database, you will need to provide your own databases. You can check out\nthe documentation of kraken. The toydb here above is shipped with sequana and\nshould work for demo. See sequana_taxonomy standalone for more help and\ninformation. You can also checkout the sequana documentation (kraken module).\n\n\nThe Kraken final report and blast analysis (if set) will need a taxonomic file\nstored in the sequana config directory (HOME/.config/sequana/taxonomy.dat). If\nnot already done, type this command::\n\n    sequana_multitax --update-taxonomy\n\nYou may need to call this command from time to time if unknown taxon appears in\nthe HTML reports.\n\n\nThis creates a directory with the pipeline and configuration file. You will then need \nto execute the pipeline::\n\n    cd multitax\n    sh multitax.sh  # for a local run\n\nThis launch a snakemake pipeline. If you are familiar with snakemake, you can \nretrieve the pipeline itself and its configuration files and then execute the pipeline yourself with specific parameters::\n\n    snakemake -s multitax.rules -c config.yaml --cores 4 --stats stats.txt\n\nOr use `sequanix <https://sequana.readthedocs.io/en/main/sequanix.html>`_ interface.\n\n\nUsage with singularity::\n~~~~~~~~~~~~~~~~~~~~~~~~~\n\nWith singularity, initiate the working directory as follows::\n\n    sequana_multitax --use-singularity ...\n\nImages are downloaded in the working directory but you can store then in a directory globally (e.g.)::\n\n    sequana_multitax --use-singularity --singularity-prefix ~/.sequana/apptainers\n\nand then::\n\n    cd multitax\n    sh multitax.sh\n\nif you decide to use snakemake manually, do not forget to add singularity options::\n\n    snakemake -s multitax.rules -c config.yaml --cores 4 --stats stats.txt --use-singularity --singularity-prefix ~/.sequana/apptainers --singularity-args "-B /home:/home"\n\n\nRequirements\n~~~~~~~~~~~~\n\nThis pipelines requires the following executable(s):\n\n- kraken2\n- sequana_taxonomy\n- krona\n\n\n.. image:: https://raw.githubusercontent.com/sequana/multitax/main/sequana_pipelines/multitax/dag.png\n\nYou can download databases from kraken website. We provide some databases on\nhttps://github.com/sequana/resources. You can download a toy database as follows::\n\n    sequana_taxonomy --download toydb\n\nThe first time, a taxonomic database will be downloaded and stored locally in\n.config/sequana/taxonomy.dat file. You can update it from time to time using::\n\n    sequana_taxonomy --update-taxonomy\n\n\nDetails\n~~~~~~~~~\n\nThis pipeline runs **sequana_taxonomy** (based on kraken) in parallel on the input fastq files (paired or not). \nA brief sequana summary report is also produced. For each sample, a HTML page is\nreported with the following kind of image. This pie chart is a static image\nsummarizing the species found in your sample. Unclassified reads are in grey.\nColors correspond to a kingdom (green for viruses). If you click on the image,\nyou will be redirect to a more precise pie chart base on Krona pie chart, which\nis more interactive.\n\n.. image:: https://raw.githubusercontent.com/sequana/multitax/main/doc/images/piechart.png\n\nThe analysis is enterily based on Kraken tool. If several databases are\nprovided, they are run sequentially. This requires a careful interpretation of\nthe results. Indeed analysing data with viruses then bacteria may give different\nresults as compare to analysing with bacteria then viruses. \n\n\nRules and configuration details\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\nHere is the `latest documented configuration file <https://raw.githubusercontent.com/sequana/multitax/main/sequana_pipelines/multitax/config.yaml>`_\nto be used with the pipeline. Each rule used in the pipeline may have a section in the configuration file. \n\nChangelog\n~~~~~~~~~\n\n========= ====================================================================\nVersion   Description\n========= ====================================================================\n0.12.0    * Refactor to use new Click framework\n0.11.1    * add missing import in the main script\n          * add wrapper version in config\n0.11.0    * use latest wrappers and graphivz apptainer\n          * create and use a sequana-wrappers for the sequana_taxonomy ruke\n          * fix type when downloading taxonomy.dat\n0.10.2    * add singularity containers\n0.10.1    * fix blast run when no taxid is found and HTML report\n0.10.0    * uses new sequana wrappers and framework\n          * add ability to run blast on unclassified reads\n          * handle case of empty FastQ files\n0.9.2     * add --update-taxonomy DB option\n          * add --store-unclassified option\n0.9.1     * fix a logger issue \n0.9.0     * fix plot summary dbs (sample names). Add options in schema+config\n            file to tune the image if required.\n          * HTML now includes links towards data that generates the top plots\n          * fix case where zero sequences are found\n          * check existence of input databases\n          * add the --run argument\n          * add multitax version in the header\n          * add search box (Sequana feature) in the CSV tables\n0.8.7     * Update HTML report: fix the title of images. include table with DB\n            proportion. Text to explain images and reports\n0.8.6     * A better report with new features from sequana.taxonomy\n0.8.5     * fix typo in doc, factorise multiqc rule\n0.8.4     * implement the --from-project option\n0.8.3     * add the confidence option in sequana_taxonomy rule\n          * improve html report\n          * uses new sequana framework to speed up --help calls\n0.8.2     * less stringent on requirements (mode warning)  \n          * fix input of the multiqc rule\n0.8.1     Fix requirements.\n0.8.0     **First release.**\n========= ====================================================================\n\n\n',
    'author': 'Sequana Team',
    'author_email': 'None',
    'maintainer': 'None',
    'maintainer_email': 'None',
    'url': 'https://github.com/sequana/multitax',
    'packages': packages,
    'package_data': package_data,
    'install_requires': install_requires,
    'entry_points': entry_points,
    'python_requires': '>=3.8,<4.0',
}


setup(**setup_kwargs)
